diff -urw acl-2.2.52.src.tar.gz.orig/include/misc.h acl-2.2.52.src.tar.gz/include/misc.h
--- acl-2.2.52.src.tar.gz.orig/include/misc.h	2013-04-30 19:07:09.000000000 +0400
+++ acl-2.2.52.src.tar.gz/include/misc.h	2015-02-05 18:29:25.100000000 +0300
@@ -17,7 +17,38 @@
 
 extern int high_water_alloc(void **buf, size_t *bufsize, size_t newsize);
 
-extern const char *quote(const char *str, const char *quote_chars);
+static inline const char *quote(const char *str, const char *quote_chars) {
+        static char *quoted_str;
+        static size_t quoted_str_len;
+        const unsigned char *s;
+        char *q;
+        size_t nonpr;
+
+        if (!str)
+                return str;
+
+        for (nonpr = 0, s = (unsigned char *)str; *s != '\0'; s++)
+                if (*s == '\\' || strchr(quote_chars, *s))
+                        nonpr++;
+        if (nonpr == 0)
+                return str;
+
+        if (high_water_alloc((void **)&quoted_str, &quoted_str_len,
+                             (s - (unsigned char *)str) + nonpr * 3 + 1))
+                return NULL;
+        for (s = (unsigned char *)str, q = quoted_str; *s != '\0'; s++) {
+                if (*s == '\\' || strchr(quote_chars, *s)) {
+                        *q++ = '\\';
+                        *q++ = '0' + ((*s >> 6)    );
+                        *q++ = '0' + ((*s >> 3) & 7);
+                        *q++ = '0' + ((*s     ) & 7);
+                } else
+                        *q++ = *s;
+        }
+        *q++ = '\0';
+
+        return quoted_str;
+}
 extern char *unquote(char *str);
 
 extern char *next_line(FILE *file);
diff -urw acl-2.2.52.src.tar.gz.orig/libmisc/quote.c acl-2.2.52.src.tar.gz/libmisc/quote.c
--- acl-2.2.52.src.tar.gz.orig/libmisc/quote.c	2013-05-19 05:58:24.000000000 +0400
+++ acl-2.2.52.src.tar.gz/libmisc/quote.c	2015-02-05 18:30:39.672000000 +0300
@@ -23,36 +23,3 @@
 #include <string.h>
 #include "misc.h"
 
-const char *quote(const char *str, const char *quote_chars)
-{
-	static char *quoted_str;
-	static size_t quoted_str_len;
-	const unsigned char *s;
-	char *q;
-	size_t nonpr;
-
-	if (!str)
-		return str;
-
-	for (nonpr = 0, s = (unsigned char *)str; *s != '\0'; s++)
-		if (*s == '\\' || strchr(quote_chars, *s))
-			nonpr++;
-	if (nonpr == 0)
-		return str;
-
-	if (high_water_alloc((void **)&quoted_str, &quoted_str_len,
-			     (s - (unsigned char *)str) + nonpr * 3 + 1))
-		return NULL;
-	for (s = (unsigned char *)str, q = quoted_str; *s != '\0'; s++) {
-		if (*s == '\\' || strchr(quote_chars, *s)) {
-			*q++ = '\\';
-			*q++ = '0' + ((*s >> 6)    );
-			*q++ = '0' + ((*s >> 3) & 7);
-			*q++ = '0' + ((*s     ) & 7);
-		} else
-			*q++ = *s;
-	}
-	*q++ = '\0';
-
-	return quoted_str;
-}
