diff -urw mesa-10.5.3-orig/configure.ac mesa-10.5.3/configure.ac
--- mesa-10.5.3-orig/configure.ac	2015-04-09 14:28:58.000000000 +0300
+++ mesa-10.5.3/configure.ac	2015-09-14 23:54:29.324000000 +0300
@@ -1938,6 +1938,11 @@
 }
 
 radeon_llvm_check() {
+    if test ${LLVM_VERSION_INT} -lt 307; then
+        amdgpu_llvm_target_name='r600'
+    else
+        amdgpu_llvm_target_name='amdgpu'
+    fi
     if test "x$enable_gallium_llvm" != "xyes"; then
         AC_MSG_ERROR([--enable-gallium-llvm is required when building $1])
     fi
@@ -1947,12 +1952,10 @@
     if test "${LLVM_VERSION_INT}${LLVM_VERSION_PATCH}" -lt "${LLVM_REQUIRED_VERSION_MAJOR}0${LLVM_REQUIRED_VERSION_MINOR}${LLVM_REQUIRED_VERSION_PATCH}"; then
         AC_MSG_ERROR([LLVM $LLVM_REQUIRED_VERSION_MAJOR.$LLVM_REQUIRED_VERSION_MINOR.$LLVM_REQUIRED_VERSION_PATCH or newer is required for $1])
     fi
-    if test true && $LLVM_CONFIG --targets-built | grep -qvw 'R600' ; then
-        AC_MSG_ERROR([LLVM R600 Target not enabled.  You can enable it when building the LLVM
-                      sources with the --enable-experimental-targets=R600
-                      configure flag])
+    if test true && $LLVM_CONFIG --targets-built | grep -iqvw $amdgpu_llvm_target_name ; then
+        AC_MSG_ERROR([LLVM $amdgpu_llvm_target_name not enabled in your LLVM build.])
     fi
-    LLVM_COMPONENTS="${LLVM_COMPONENTS} r600 bitreader ipo"
+    LLVM_COMPONENTS="${LLVM_COMPONENTS} $amdgpu_llvm_target_name bitreader ipo"
     NEED_RADEON_LLVM=yes
     if test "x$have_libelf" != xyes; then
        AC_MSG_ERROR([$1 requires libelf when using llvm])
diff -urw mesa-10.5.3-orig/src/gallium/auxiliary/gallivm/lp_bld_debug.cpp mesa-10.5.3/src/gallium/auxiliary/gallivm/lp_bld_debug.cpp
--- mesa-10.5.3-orig/src/gallium/auxiliary/gallivm/lp_bld_debug.cpp	2015-04-09 14:26:33.000000000 +0300
+++ mesa-10.5.3/src/gallium/auxiliary/gallivm/lp_bld_debug.cpp	2015-09-14 23:56:14.536000000 +0300
@@ -271,9 +271,11 @@
 #if defined(PIPE_ARCH_X86)
    options.StackAlignmentOverride = 4;
 #endif
+#if HAVE_LLVM < 0x0307
 #if defined(DEBUG) || defined(PROFILE)
    options.NoFramePointerElim = true;
 #endif
+#endif
    OwningPtr<TargetMachine> TM(T->createTargetMachine(Triple, sys::getHostCPUName(), "", options));
 
 #if HAVE_LLVM >= 0x0306
diff -urw mesa-10.5.3-orig/src/gallium/auxiliary/gallivm/lp_bld_misc.cpp mesa-10.5.3/src/gallium/auxiliary/gallivm/lp_bld_misc.cpp
--- mesa-10.5.3-orig/src/gallium/auxiliary/gallivm/lp_bld_misc.cpp	2015-04-09 14:26:33.000000000 +0300
+++ mesa-10.5.3/src/gallium/auxiliary/gallivm/lp_bld_misc.cpp	2015-09-14 23:56:44.588000000 +0300
@@ -439,8 +439,10 @@
 #if HAVE_LLVM < 0x0304
    options.NoFramePointerElimNonLeaf = true;
 #endif
+#if HAVE_LLVM < 0x0307
    options.NoFramePointerElim = true;
 #endif
+#endif
 
    builder.setEngineKind(EngineKind::JIT)
           .setErrorStr(&Error)
diff -urw mesa-10.5.3-orig/src/gallium/drivers/radeon/radeon_llvm_emit.c mesa-10.5.3/src/gallium/drivers/radeon/radeon_llvm_emit.c
--- mesa-10.5.3-orig/src/gallium/drivers/radeon/radeon_llvm_emit.c	2015-03-21 03:51:18.000000000 +0300
+++ mesa-10.5.3/src/gallium/drivers/radeon/radeon_llvm_emit.c	2015-09-14 23:55:34.412000000 +0300
@@ -86,10 +86,17 @@
 {
 	static unsigned initialized = 0;
 	if (!initialized) {
+#if HAVE_LLVM < 0x0307
 		LLVMInitializeR600TargetInfo();
 		LLVMInitializeR600Target();
 		LLVMInitializeR600TargetMC();
 		LLVMInitializeR600AsmPrinter();
+#else
+		LLVMInitializeAMDGPUTargetInfo();
+		LLVMInitializeAMDGPUTarget();
+		LLVMInitializeAMDGPUTargetMC();
+		LLVMInitializeAMDGPUAsmPrinter();
+#endif
 		initialized = 1;
 	}
 }
